<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Vignette for clipp</title>

<script src="data:application/javascript;base64,Ly8gUGFuZG9jIDIuOSBhZGRzIGF0dHJpYnV0ZXMgb24gYm90aCBoZWFkZXIgYW5kIGRpdi4gV2UgcmVtb3ZlIHRoZSBmb3JtZXIgKHRvCi8vIGJlIGNvbXBhdGlibGUgd2l0aCB0aGUgYmVoYXZpb3Igb2YgUGFuZG9jIDwgMi44KS4KZG9jdW1lbnQuYWRkRXZlbnRMaXN0ZW5lcignRE9NQ29udGVudExvYWRlZCcsIGZ1bmN0aW9uKGUpIHsKICB2YXIgaHMgPSBkb2N1bWVudC5xdWVyeVNlbGVjdG9yQWxsKCJkaXYuc2VjdGlvbltjbGFzcyo9J2xldmVsJ10gPiA6Zmlyc3QtY2hpbGQiKTsKICB2YXIgaSwgaCwgYTsKICBmb3IgKGkgPSAwOyBpIDwgaHMubGVuZ3RoOyBpKyspIHsKICAgIGggPSBoc1tpXTsKICAgIGlmICghL15oWzEtNl0kL2kudGVzdChoLnRhZ05hbWUpKSBjb250aW51ZTsgIC8vIGl0IHNob3VsZCBiZSBhIGhlYWRlciBoMS1oNgogICAgYSA9IGguYXR0cmlidXRlczsKICAgIHdoaWxlIChhLmxlbmd0aCA+IDApIGgucmVtb3ZlQXR0cmlidXRlKGFbMF0ubmFtZSk7CiAgfQp9KTsK"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="data:text/css,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Vignette for clipp</h1>



<p>The R package <code>clipp</code> provides a fast and general implementation of the Elston-Stewart algorithm.</p>
<p>The main function is <code>pedigree_loglikelihood</code>, which calculates the pedigree log-likelihood for almost any choice of genetic model. Helper functions are provided that specify commonly used genetic models, though users are free to define and use more advanced ones. Combining <code>clipp</code> with an optimisation function like <code>mle</code> allows the user to perform maximum-likelihood estimation of model parameters, as illustrated below.</p>
<p><code>clipp</code> also provides a function, <code>genotype_probabilities</code>, that calculates genotype probabilities for a target person within a family, given the family’s observed data.</p>
<div id="pedigree-likelihoods" class="section level2">
<h2>Pedigree likelihoods</h2>
<p>The function <code>pedigree_loglikelihood</code> calculates (the logarithm of) any pedigree likelihood that can be written in Ott’s form, as given on page 117 of (Lange, 2002). In this formulation, the likelihood <span class="math inline">\(L\)</span> for a given family is <span class="math display">\[ L = \sum_{g_1} \dots \sum_{g_n} \prod_{i=1}^n P(x_i \mid g_i) P(g_i \mid g_{m(i)}, g_{f(i)}),\]</span> where: there are <span class="math inline">\(n\)</span> members of the family, who are labeled by identifiers <span class="math inline">\(i = 1, \dots, n\)</span>; <span class="math inline">\(g_i\)</span> and <span class="math inline">\(x_i\)</span> are the genotype and observed data for person <span class="math inline">\(i\)</span>, respectively; <span class="math inline">\(P(x_i \mid g_i)\)</span> is the conditional probability of person <span class="math inline">\(i\)</span>’s observed data give his or her genotype; <span class="math inline">\(m(i)\)</span> and <span class="math inline">\(f(i)\)</span> are the identifiers of person <span class="math inline">\(i\)</span>’s mother and father, respectively, or missing if person <span class="math inline">\(i\)</span> is a founder (i.e. if person <span class="math inline">\(i\)</span> does not have parents in the pedigree); and <span class="math inline">\(P(g_i \mid g_{m(i)}, g_{f(i)})\)</span> is either the population prevalence of genotype <span class="math inline">\(g_i\)</span> (if person <span class="math inline">\(i\)</span> is a founder) or the conditional probability of offspring genotype <span class="math inline">\(g_i\)</span> given parental genotypes <span class="math inline">\(g_{m(i)}\)</span> and <span class="math inline">\(g_{f(i)}\)</span> (if person <span class="math inline">\(i\)</span> is a non-founder). Each sum is over all possible genotypes at the genetic locus under consideration (or over all possible multi-locus genotypes if we are modeling two or more genetic loci).</p>
<p>The terms in this likelihood correspond to the main arguments of the function <code>pedigree_loglikelihood</code>. The argument <code>dat</code> specifies the family structure, i.e. <span class="math inline">\(m(i)\)</span> and <span class="math inline">\(f(i)\)</span> for all family members <span class="math inline">\(i\)</span>. The argument <code>geno_freq</code> specifies the population genotype prevalences, i.e.  <span class="math inline">\(P(g_i \mid g_{m(i)}, g_{f(i)})\)</span> when person <span class="math inline">\(i\)</span> is a founder. The argument <code>trans</code> specifies the transmission probabilities, i.e. <span class="math inline">\(P(g_i \mid g_{m(i)}, g_{f(i)})\)</span> when person <span class="math inline">\(i\)</span> is a non-founder. Lastly, the penetrance matrix <code>penet</code> specifies the relationship between the observed data and the genotypes, i.e. <span class="math inline">\(P(x_i \mid g_i)\)</span>.</p>
</div>
<div id="specifying-the-genetic-model" class="section level2">
<h2>Specifying the genetic model</h2>
<p>The transmission probabilities <code>trans</code> and population genotype prevalences <code>geno_freq</code> determine the joint probability for any combination of genotypes within a family, so we say that <code>trans</code> and <code>geno_freq</code> define the genetic model of the analysis. These terms are usually based on known genetic laws (such as Mendel’s laws) and user-specified genotype or allele frequencies. <code>clipp</code> provides helper functions to calculate <code>trans</code> and <code>geno_freq</code> for common genetic models. For example, the following code specifies a genetic model consisting of a single biallelic locus with a minor allele frequency of 10%.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(clipp)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>MAF <span class="ot">&lt;-</span> <span class="fl">0.1</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>geno_freq <span class="ot">&lt;-</span> <span class="fu">geno_freq_monogenic</span>(<span class="at">p_alleles =</span> <span class="fu">c</span>(<span class="dv">1</span> <span class="sc">-</span> MAF, MAF))</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>trans <span class="ot">&lt;-</span> <span class="fu">trans_monogenic</span>(<span class="at">n_alleles =</span> <span class="dv">2</span>)</span></code></pre></div>
<p>We can view a more user-friendly version of the genotype frequencies by using the option <code>annotate = TRUE</code>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">geno_freq_monogenic</span>(<span class="at">p_alleles =</span> <span class="fu">c</span>(<span class="dv">1</span> <span class="sc">-</span> MAF, MAF), <span class="at">annotate =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  1/1  1/2  2/2 </span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 0.81 0.18 0.01</span></span></code></pre></div>
<p>This shows that the alleles at the genetic locus have been given the default names <code>1</code> and <code>2</code>, and that the possible genotypes are <code>1/1</code>, <code>1/2</code> and <code>2/2</code>. In this example, the population prevalence of the heterozygous genotype <code>1/2</code> is 18%.</p>
<p>The option <code>annotate = TRUE</code> also reveals a more user-friendly version of the transmission probabilities.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">trans_monogenic</span>(<span class="at">n_alleles =</span> <span class="dv">2</span>, <span class="at">annotate =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    gm  gf  1/1 1/2  2/2</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 1/1 1/1 1.00 0.0 0.00</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2 1/1 1/2 0.50 0.5 0.00</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3 1/1 2/2 0.00 1.0 0.00</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4 1/2 1/1 0.50 0.5 0.00</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5 1/2 1/2 0.25 0.5 0.25</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 6 1/2 2/2 0.00 0.5 0.50</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 7 2/2 1/1 0.00 1.0 0.00</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 8 2/2 1/2 0.00 0.5 0.50</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 9 2/2 2/2 0.00 0.0 1.00</span></span></code></pre></div>
<p>Here, the rows correspond to the nine possible joint parental genotypes and the last three columns correspond to the three possible offspring genotypes. Each number is the conditional probability of the offspring genotype, given the parental genotypes. For example, row <code>3</code> says that when the mother has genotype <code>1/1</code> and the father has genotype <code>2/2</code> then the offspring will have genotype <code>1/2</code> (with probability <code>1</code>). The probabilities in this table are given by Mendel’s laws of genetics. Note that the probabilities in each row sum to <code>1</code>, since the offspring must always have one of the three possible genotypes, regardless of the parental genotypes.</p>
</div>
<div id="calculating-the-pedigree-likelihood" class="section level2">
<h2>Calculating the pedigree likelihood</h2>
<p>Now that we’ve defined a genetic model, we can use the function <code>pedigree_loglikelihood</code> to calculate the pedigree log-likelihoods of some sample families.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;dat_small&quot;</span>, <span class="st">&quot;penet_small&quot;</span>, <span class="st">&quot;dat_large&quot;</span>, <span class="st">&quot;penet_large&quot;</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dat_small)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   family  indiv mother father sex aff age geno</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1    ora ora001 ora009 ora010   1   1  42  2/2</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2    ora ora002   &lt;NA&gt;   &lt;NA&gt;   2   0   6  1/1</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3    ora ora003 ora002 ora001   1   0  22     </span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4    ora ora004 ora002 ora001   2   0  17  1/2</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5    ora ora005 ora002 ora001   1   0  22     </span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 6    ora ora006 ora002 ora001   1   0  71</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(penet_small)</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;           [,1]      [,2]      [,3]</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1,] 0.9988662 0.9966026 0.9966026</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [2,] 0.9999994 0.9999982 0.9999982</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [3,] 0.9999676 0.9999028 0.9999028</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [4,] 0.9999866 0.9999598 0.9999598</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [5,] 0.9999676 0.9999028 0.9999028</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [6,] 0.9561271 0.8740714 0.8740714</span></span></code></pre></div>
<p>Each row of <code>dat_small</code> corresponds to a person in one of 10 families. Here, <code>dat_small</code> specifies the family structure for these 10 families, meaning that <code>dat_small</code> specifies the mother and father of each person in each family. Some people, such as person <code>ora002</code>, have missing parental identifiers because they are founders (i.e. their parents are not included in the pedigree). Each person has either both or no parents included in their pedigree, i.e. if the mother identifier is missing then so is the father identifier, and vice versa. Also, each person who is mentioned as a parent has a corresponding row, e.g.  person <code>ora009</code> is listed as the mother of person <code>ora001</code>, so there must be a row later in the pedigree with <code>indiv</code> equal to <code>ora009</code>. The columns <code>sex</code>, <code>aff</code>, <code>age</code> and <code>geno</code> are ignored by the function <code>pedigree_loglikelihood</code>, though data like this will usually contribute to the corresponding pedigree likelihood via the penetrance matrix. We will soon give examples of calculating penetrance matrices from this sort of observed data, but for now we simply use the sample penetrance matrix <code>penet_small</code>.</p>
<p>If there are any identical twins or triplets in the family then we can specify them using an optional argument <code>monozyg</code>. For example, to indicate that <code>ora024</code> and <code>ora027</code> are identical twins, and so are <code>aey063</code> and <code>aey064</code>, then we can use the following as the <code>monozyg</code> argument:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>monozyg_small <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="st">&quot;ora024&quot;</span>, <span class="st">&quot;ora027&quot;</span>), <span class="fu">c</span>(<span class="st">&quot;aey063&quot;</span>, <span class="st">&quot;aey064&quot;</span>))</span></code></pre></div>
<p>We can now calculate the log-likelihoods of the 10 families.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pedigree_loglikelihood</span>(dat_small, geno_freq, trans, penet_small, </span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">monozyg =</span> monozyg_small, <span class="at">sum_loglik =</span> <span class="cn">FALSE</span>, <span class="at">ncores =</span> <span class="dv">2</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;       ora       ltk       aey       fwj       tqo       ibz       shd       vtc </span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -224.0860 -170.4035 -204.6862 -198.4982 -193.2291 -179.5359 -246.8078 -160.9817 </span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;       jil       mak </span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -163.6852 -158.1138</span></span></code></pre></div>
<p>By default, <code>pedigree_loglikelihood</code> will return the sum of the pedigree log-likelihoods of all of the families. However, the option <code>sum_loglik = FALSE</code> above instructs <code>pedigree_loglikelihood</code> to return the separate log-likelihoods, e.g.  the above output shows that family <code>ora</code> has a log-likelihood of <code>-224.0860</code>. The option <code>ncores = 2</code> instructs <code>pedigree_loglikelihood</code> to perform this calculation in parallel on two cores, with the log-likelihoods of five families calculated on one core and the remaining five on another core.</p>
<p>The function <code>pedigree_loglikelihood</code> can also handle very large families. For example, the following code takes much less than one minute on a standard desktop computer to calculate the log-likelihood of a family with approximately 10,000 family members.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">system.time</span>(ll <span class="ot">&lt;-</span> <span class="fu">pedigree_loglikelihood</span>(dat_large, geno_freq, trans, penet_large))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   10.64    0.15   10.83</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>ll</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] -18020.99</span></span></code></pre></div>
</div>
<div id="maximum-likelihood-estimation" class="section level2">
<h2>Maximum likelihood estimation</h2>
<p>The argument <code>penet</code> of <code>pedigree_loglikelihood</code> specifies the penetrance matrix, which usually depends on the observed data and some unknown parameters that we would like to estimate. In this section, we give a simple example of such a penetrance matrix and then use this to estimate its unknown parameters.</p>
<p>Let’s take <code>dat_small</code> as our sample data, and suppose we’re interested in estimating the log-odds of disease for the three possible genotypes from previous sections. Here, the log-odds of disease is <code>log(p/(1-p))</code> when <code>p</code> is the probability of disease.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dat_small)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   family  indiv mother father sex aff age geno</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1    ora ora001 ora009 ora010   1   1  42  2/2</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2    ora ora002   &lt;NA&gt;   &lt;NA&gt;   2   0   6  1/1</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3    ora ora003 ora002 ora001   1   0  22     </span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4    ora ora004 ora002 ora001   2   0  17  1/2</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5    ora ora005 ora002 ora001   1   0  22     </span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 6    ora ora006 ora002 ora001   1   0  71</span></span></code></pre></div>
<p>As a toy example, we ignore age and most other data, and take the observed data to be just the disease affected status <code>aff</code>. For simplicity, we also assume that allele <code>1</code> is dominant to allele <code>2</code>, meaning that the log-odds of disease are the same for people with genotypes <code>1/1</code> and <code>1/2</code>. The parameters that we would like to estimate are therefore the log-odds of disease for genotypes <code>1/1</code> and <code>1/2</code> (combined) and for genotype <code>2/2</code>. We denote these parameters by <code>logodds1</code> and <code>logodds2</code>, respectively.</p>
<p>A penetrance matrix <code>penet</code> is designed to give a probabilistic connection between a peron’s genotype and his or her observed data. Each row of <code>penet</code> corresponds to a person and each column corresponds to one of the possible genotypes. (Row <code>i</code> of <code>penet</code> and row <code>i</code> of <code>dat</code> should correspond to the same person, and the order of the genotypes should be the same for <code>penet</code>, <code>geno_freq</code> and <code>trans</code>.) The <code>(i, j)</code><sup>th</sup> entry of <code>penet</code> gives the conditional probability of person <code>i</code>’s observed data, given that he or she has the <code>j</code><sup>th</sup> genotype. In our example, these conditional probabilities are determined by the parameters <code>logodds1</code> and <code>logodds2</code>. For instance, if person <code>i</code> has genotype <code>1/1</code> then his or her probability of disease is <code>prob1 = 1/(1 + exp(-logodds1))</code>, so if person <code>i</code> is affected then the conditional probability of his or her observed data is <code>prob1</code>, and if person <code>i</code> is unaffected then this conditional probability is <code>1 - prob1</code>. Therefore, the <code>(i, 1)</code><sup>th</sup> entry of <code>penet</code>, corresponding to person <code>i</code> and the first genotype (i.e. <code>1/1</code>), is <code>prob1</code> if person <code>i</code> is affected and <code>1 - prob1</code> if person <code>i</code> is unaffected. The following function uses this and similar reasoning to calculate row <code>i</code> of <code>penet</code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>penet.fn <span class="ot">&lt;-</span> <span class="cf">function</span>(i, logodds1, logodds2) {</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  prob1 <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>(<span class="dv">1</span> <span class="sc">+</span> <span class="fu">exp</span>(<span class="sc">-</span>logodds1))</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  prob2 <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>(<span class="dv">1</span> <span class="sc">+</span> <span class="fu">exp</span>(<span class="sc">-</span>logodds2))</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  penet.i <span class="ot">&lt;-</span> <span class="fu">c</span>(prob1, prob1, prob2)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (dat_small<span class="sc">$</span>aff[i] <span class="sc">==</span> <span class="dv">0</span>)  penet.i <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> penet.i</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(penet.i)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>We can now estimate our parameters of interest, <code>logodds1</code> and <code>logodds2</code>, using maximum likelihood estimation, as implemented in the <code>mle</code> function of the <code>stats4</code> package.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stats4)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>minusll <span class="ot">&lt;-</span> <span class="cf">function</span>(<span class="at">logodds1 =</span> <span class="dv">0</span>, <span class="at">logodds2 =</span> <span class="dv">0</span>) {</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  penet <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(dat_small), penet.fn, logodds1, logodds2))</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  loglik <span class="ot">&lt;-</span> <span class="fu">pedigree_loglikelihood</span>(dat_small, geno_freq, trans, penet,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">monozyg =</span> monozyg_small, <span class="at">ncores =</span> <span class="dv">2</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="sc">-</span>loglik)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="fu">minusll</span>()</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 705.6238</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">mle</span>(minusll)</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit)</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Maximum likelihood estimation</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; mle(minuslogl = minusll)</span></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Coefficients:</span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;           Estimate Std. Error</span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; logodds1 -1.359962  0.1054336</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; logodds2 -1.313467  6.8597364</span></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -2 log L: 1030.901</span></span></code></pre></div>
<p>The function <code>minusll</code> above first calculates the penetrance matrix <code>penet</code> corresponding to any given values of the parameters <code>logodds1</code> and <code>logodds2</code>, and then uses this matrix and <code>pedigree_loglikelihood</code> to calculate the corresponding log-likelihood. The above code then uses <code>mle</code> to perform maximum likelihood estimation, giving estimates of <code>-1.36</code> and <code>-1.31</code> for <code>logodds1</code> and <code>logodds2</code>, respectively.</p>
</div>
<div id="incorporating-known-genotypes" class="section level2">
<h2>Incorporating known genotypes</h2>
<p>We will now extend the analysis of the previous section to include known genotypes.</p>
<p>The dataset <code>dat_small</code> has a column <code>geno</code> corresponding to known genotypes. As is often the case with family data, many people in <code>dat_small</code> do not have measured genotypes, and so have a blank (<code>&quot;&quot;</code>) in the <code>geno</code> column.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dat_small)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   family  indiv mother father sex aff age geno</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1    ora ora001 ora009 ora010   1   1  42  2/2</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2    ora ora002   &lt;NA&gt;   &lt;NA&gt;   2   0   6  1/1</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3    ora ora003 ora002 ora001   1   0  22     </span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4    ora ora004 ora002 ora001   2   0  17  1/2</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5    ora ora005 ora002 ora001   1   0  22     </span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 6    ora ora006 ora002 ora001   1   0  71</span></span></code></pre></div>
<p>The standard method for incorporating known genotypes into the Elston-Stewart algorithm is to first calculate the penetrance matrix <code>penet</code> while ignoring all genotype data, and then change <code>penet</code> for people who have measured genotypes according to the following simple rule. If person <code>i</code> is known to have the <code>j</code><sup>th</sup> genotype then <code>penet[i, j]</code> is unchanged but <code>penet[i, k]</code> is set to <code>0</code> for all <code>k != j</code>. See later for brief justification for this rule.</p>
<p>For example, to incorporate known genotypes into the analysis of the previous section, we modify <code>penet.fn</code> by adding the three lines of code below that are marked with <span class="math inline">\(\color{darkred}{\text{###}}\)</span>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>penet.fn <span class="ot">&lt;-</span> <span class="cf">function</span>(i, logodds1, logodds2) {</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  prob1 <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>(<span class="dv">1</span> <span class="sc">+</span> <span class="fu">exp</span>(<span class="sc">-</span>logodds1))</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  prob2 <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>(<span class="dv">1</span> <span class="sc">+</span> <span class="fu">exp</span>(<span class="sc">-</span>logodds2))</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  penet.i <span class="ot">&lt;-</span> <span class="fu">c</span>(prob1, prob1, prob2)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (dat_small<span class="sc">$</span>aff[i] <span class="sc">==</span> <span class="dv">0</span>)  penet.i <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> penet.i</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (dat_small<span class="sc">$</span>geno[i] <span class="sc">==</span> <span class="st">&quot;1/1&quot;</span>)  penet.i[<span class="sc">-</span><span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>     <span class="do">###</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (dat_small<span class="sc">$</span>geno[i] <span class="sc">==</span> <span class="st">&quot;1/2&quot;</span>)  penet.i[<span class="sc">-</span><span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>     <span class="do">###</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (dat_small<span class="sc">$</span>geno[i] <span class="sc">==</span> <span class="st">&quot;2/2&quot;</span>)  penet.i[<span class="sc">-</span><span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>     <span class="do">###</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(penet.i)</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>We can now use the same code as in the previous section to estimate the log-odds of disease, though our estimates are now based on the known genotypes as well as disease affected statuses.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">minusll</span>()</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 788.5003</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">mle</span>(minusll)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Maximum likelihood estimation</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; mle(minuslogl = minusll)</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Coefficients:</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;           Estimate Std. Error</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; logodds1 -1.357596 0.08405912</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; logodds2 -1.395321 0.61632248</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; -2 log L: 1196.65</span></span></code></pre></div>
<p>The above method for incorporating known genotypes can be justified as follows. Measured genotypes can differ from actual genotypes due to genotyping errors, and even when genotyping errors are negligible, we can make a conceptual distinction between the actual genotype <span class="math inline">\(g_i\)</span> of person <code>i</code>, and any measured genotype of person <code>i</code>, which is part of his or her observed data <span class="math inline">\(x_i\)</span>. The <code>(i, k)</code><sup>th</sup> component <code>penet[i, k]</code> of the penetrance matrix is the conditional probability of person <code>i</code>’s observed data <span class="math inline">\(x_i\)</span>, given that his or her actual genotype <span class="math inline">\(g_i\)</span> is the <code>k</code><sup>th</sup> genotype (e.g. <code>2/2</code> is the third genotype, in our running example). When genotyping errors are negligible, the chance of observing a genotype different from the actual genotype is <code>0</code>. So if person <code>i</code> is observed to have the <code>j</code><sup>th</sup> genotype then <code>penet[i, k]</code> is <code>0</code> for all <code>k != j</code>. Using similar reasoning, non-negligible genotyping errors and partial genotyping information can also be incorporated in an analysis.</p>
</div>
<div id="calculating-carrier-probabilities" class="section level2">
<h2>Calculating carrier probabilities</h2>
<p>Given a genetic model and a penetrance matrix, <code>clipp</code> can also use the Elston-Stewart algorithm to calculate genotype probabilities for a person of interest, using the function <code>genotype_probabilities</code>. For example, we can calculate the genotype probabilities for individual <code>ora008</code> in the family <code>ora</code>, as follows.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dat_small)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   family  indiv mother father sex aff age geno</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1    ora ora001 ora009 ora010   1   1  42  2/2</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2    ora ora002   &lt;NA&gt;   &lt;NA&gt;   2   0   6  1/1</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3    ora ora003 ora002 ora001   1   0  22     </span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4    ora ora004 ora002 ora001   2   0  17  1/2</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5    ora ora005 ora002 ora001   1   0  22     </span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 6    ora ora006 ora002 ora001   1   0  71</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="fu">genotype_probabilities</span>(<span class="at">target =</span> <span class="st">&quot;ora008&quot;</span>, dat_small, geno_freq, trans, </span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>                       penet_small, monozyg_small)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.19832320 0.74111557 0.06056123</span></span></code></pre></div>
<p>This shows that person <code>ora008</code> has a <code>19.8%</code> chance of having the first genotype (<code>1/1</code>), given the inputs (family structure, genetic model and penetrance matrix). In this calculation, the penetrance matrix <code>penet_small</code> depends on all of the observed data. If we instead wanted genotype probabilities that are based only on the family structure and observed genotypes then we could use the following code.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>penet.fn <span class="ot">&lt;-</span> <span class="cf">function</span>(i) {</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  penet.i <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">3</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (dat_small<span class="sc">$</span>geno[i] <span class="sc">==</span> <span class="st">&quot;1/1&quot;</span>)  penet.i[<span class="sc">-</span><span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (dat_small<span class="sc">$</span>geno[i] <span class="sc">==</span> <span class="st">&quot;1/2&quot;</span>)  penet.i[<span class="sc">-</span><span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (dat_small<span class="sc">$</span>geno[i] <span class="sc">==</span> <span class="st">&quot;2/2&quot;</span>)  penet.i[<span class="sc">-</span><span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(penet.i)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>penet <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(dat_small), penet.fn))</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="fu">genotype_probabilities</span>(<span class="at">target =</span> <span class="st">&quot;ora008&quot;</span>, dat_small, geno_freq, trans, </span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>                       penet, monozyg_small)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.45 0.50 0.05</span></span></code></pre></div>
</div>
<div id="advanced-notes" class="section level2">
<h2>Advanced notes</h2>
<p>Many statistical analyses of family data can be performed by the Elston-Stewart algorithm combined with maximum likelihood estimation, for some choice of genetic model and penetrance matrix. We have given simple examples of segregation analyses to estimate disease risk. These can be modified slightly and combined with a genetic model based on phased genotypes (as given by <code>clipp</code>’s functions <code>geno_freq_phased</code> and <code>trans_phased</code>) to investigate parent-of-origin effects. With some ingenuity, the user can also use <code>clipp</code> to investigate other interesting genetic models, such as linked loci and non-standard transmission probabilities, or whatever genetic model can be imagined.</p>
<p>Ott’s form for the pedigree likelihood assumes that the observed data of different family members is conditionally independent, given their genotypes. However, this assumption can be weakened for a genetic locus of interest, by adding an unmeasured polygene into the genetic model (using functions like <code>trans_polygenic</code> and <code>combine_loci</code>).</p>
</div>
<div id="the-elston-stewart-algorithm" class="section level2">
<h2>The Elston-Stewart algorithm</h2>
<p>General references for the Elston-Stewart algorithm are (Elston &amp; Stewart, 1971), (Lange &amp; Elston, 1975) and (Cannings et al., 1978). Up to logarithmic factors, the complexity of the algorithm is <code>O(n * m)</code>, where <code>n</code> is the number of people in the family and <code>m</code> is the number of possible genotypes.</p>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<p>Cannings C, Thompson E, Skolnick M. Probability functions on complex pedigrees. Advances in Applied Probability, 1978;10(1):26-61.</p>
<p>Elston RC, Stewart J. A general model for the genetic analysis of pedigree data. Hum Hered. 1971;21(6):523-542.</p>
<p>Lange K. Mathematical and Statistical Methods for Genetic Analysis (second edition). Springer, New York. 2002.</p>
<p>Lange K, Elston RC. Extensions to pedigree analysis I. Likehood calculations for simple and complex pedigrees. Hum Hered. 1975;25(2):95-105.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
